### ISSUE TITLE
Missed testcleanup in decimal.rst

### ISSUE THREAD

**Issue by c6a24554-5737-4174-b47e-54fc0403b8dc** [python/cpython#74241]
BPO | [30055](https://bugs.python.org/issue30055)
--- | :---
Nosy | @skrah, @marco-buttu
PRs | <li>python/cpython#1094</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/skrah'
closed_at = <Date 2017-04-13.11:35:49.135>
created_at = <Date 2017-04-12.14:59:31.646>
labels = ['type-feature', '3.7', 'docs']
title = 'Missed testcleanup in decimal.rst'
updated_at = <Date 2017-04-13.12:23:50.887>
user = 'https://github.com/marco-buttu'
```

bugs.python.org fields:
```python
activity = <Date 2017-04-13.12:23:50.887>
actor = 'skrah'
assignee = 'skrah'
closed = True
closed_date = <Date 2017-04-13.11:35:49.135>
closer = 'skrah'
components = ['Documentation']
creation = <Date 2017-04-12.14:59:31.646>
creator = 'marco.buttu'
dependencies = []
files = []
hgrepos = []
issue_num = 30055
keywords = []
message_count = 5.0
messages = ['291559', '291601', '291603', '291604', '291606']
nosy_count = 3.0
nosy_names = ['skrah', 'docs@python', 'marco.buttu']
pr_nums = ['1094']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30055'
versions = ['Python 3.7']
```

</p></details>

**Comment by c6a24554-5737-4174-b47e-54fc0403b8dc** [comment]
The testsetup in [Doc/library/decimal.rst](https://github.com/python/cpython/blob/main/Doc/library/decimal.rst) is not enough for isolating the tests in respect to the other rst files.  Currently we have the following testsetup, without a testcleanup:

 .. testsetup:: *
 
    import decimal
    import math
    from decimal import *
    \# make sure each group gets a fresh context
    setcontext(Context())

Without a testcleanup, the changes on the context will affect the other files that use the context (like [Doc/library/statistics.rst](https://github.com/python/cpython/blob/main/Doc/library/statistics.rst)).  We should better isolate the tests adding also a testcleanup:

.. testcleanup:: \*

```py
   # make sure other tests (outside this file) get a fresh context
   setcontext(Context())
```


I am opening a PR.

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
New changeset 2c0b5c664ba7d36dc09788d3778d5b33e3fa1bd0 by Stefan Krah (Marco Buttu) in branch 'master':
bpo-30055: add testcleanup to leave a fresh context (bpo-1094)
https://github.com/python/cpython/commit/2c0b5c664ba7d36dc09788d3778d5b33e3fa1bd0

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Thanks, looks good.  (I hope GitHub runs the doctests at Travis CI, locally most of them failed so I didn't bother to run the tests for this one.)

**Comment by c6a24554-5737-4174-b47e-54fc0403b8dc** [comment]
We are not executing the doctests on Travis CI, but I executed them locally.  That's why I realized there was a missed testcleanup.  The bpo-27200 wants to fix all doctests, and I actually completed the job, but three PRs are still opened, and I am waiting for review approval.  If you want to contribute to the reviews, look for 'bpo-27200' on the GitHub PRs.  I hope the doctest will be run on Travis as soon as possible.  Thanks, Marco
